text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+
+The `mod-pubsub` module should be able to obtain a JWT token from mod-authtoken on behalf of that “technical” user as well. It could be done when the module is enabled for a tenant. Time-to-live for such a token should be quite long enough.
+
+A special set of FOLIO permissions should be defined to allow to publish and deliver events.
+
+#### Implementation notes
+The `pubsub` implementation (`mod-pubsub`) must have its own persistent storage to store all registered event descriptors with publishers and subscribers linked to them.
+
+A dedicated git repository can be used to store event descriptors and it must be added as a git submodule to a back-end module’s repositories. It will allow to manage event descriptors easily.
+
+## Risks and Drawbacks
+This RFC introduces the new functionality to the FOLIO platform, that leads to the increasing platform complexity.
+
+Also, it adds a dependency on the third-party platform (Kafka). But having that behind the `mod-pubsub` allows substituting messaging vendor with a minimal impact on the FOLIO platform.
+
+Some HTTP calls overhead will be present because all interactions between `mod-pubsub`, publishers and subscribers will be done via the HTTP protocol through the OKAPI module. 
+
+## Rationale and Alternatives
+It could be possible to implement `pubsub` functionality in the scope of OKAPI, but it this will unnecessarily increase the complexity of the OKAPI module, which is currently quite complex. However, such an implementation may not be sufficiently feature rich.
+
+As an alternative, it is possible to use a messaging platform (Kafka, RabbitMQ) directly without a proxy module. But it will add an explicit dependency on the selected messaging engine. 
+Also, it forces to add some boilerplate code snippets to all modules which need to deal with publisher-subscriber functionality, or at least it will require to create some sort of shared library to be used by backend modules.
+But since FOLIO is a language agnostic platform it will also have its limitations.
+  
+
+## Unresolved Questions
+
+Now the least developed area is questions related to the security, permissions, receiving JWT tokens, etc. 
```

Comment:
What might the message delivery semantics be for `mod-pubsub` for example, expected latency (for last subscriber) or delivery at least or more than once?

There is some discussion about TTL and the impact that has, I didn't notice other semantics, have I missed a section?

Reply:
I'll take the opportunity to respond here to your "broader questions" above.

The proposal here contains a simple implementation in v1. The focus is on publishing and subscribing to events - the fact that a message queue is used under the covers is merely a poorly kept secret. It is intent of this v1 proposal to expose the underlying messaging technology. (Such as guaranteed delivery or delivery at least once, etc...). Those can be added in a later version. However, doing so would definitely lock in the choice of message queue. 

Similarly, the notion of targeted messaging (as opposed to pubsub) was also discussed but is out of scope for this proposal.

Summarizing from the proposal:
- expected latency to first subscriber (TTL)
- delivery more than once (same event consumed by multiple subscribers)
- no delivery guarantee or receipts.

Reply:
> I'll take the opportunity to respond here to your "broader questions" above.
> 
> The proposal here contains a simple implementation in v1. The focus is on publishing and subscribing to events - the fact that a message queue is used under the covers is merely a poorly kept secret. It is intent of this v1 proposal to expose the underlying messaging technology. (Such as guaranteed delivery or delivery at least once, etc...). Those can be added in a later version. However, doing so would definitely lock in the choice of message queue.

Thanks for clarifying that. I only wanted to understand if we were establishing common semantics for the pubsub interface about delivery. I understand why we may want to defer that to after this initial implementation.

I appreciate that we don’t want to accidentally couple the system to the semantics of the first implementation’s underlying infrastructure.

> 
> Similarly, the notion of targeted messaging (as opposed to pubsub) was also discussed but is out of scope for this proposal.

Ok, so this should only be used for event publishing, for informing interested parties that something has occurred? 

And this mechanism should *not* be used to ask another party to do something?

I want to attune myself to this, so I can understand what appropriate use of this first version looks like.

> Summarizing from the proposal:
> 
> * expected latency to first subscriber (TTL)
> * delivery more than once (same event consumed by multiple subscribers)

And the same event can be delivered to the same subscriber more than once?

> * no delivery guarantee or receipts.

Thank you for summarising that, I likely failed to pick up on those aspects from the text directly 😕

Reply:
> Ok, so this should only be used for event publishing, for informing interested parties that something has occurred?

Correct 

> And this mechanism should not be used to ask another party to do something?

Also correct

> And the same event can be delivered to the same subscriber more than once?

Not in the initial version. Note that doing so would require imposing an idempotency requirement on the subscriber's event handler. It is probably a good requirement anyways, but I would be hesitant to impose it for the initial version as a strict requirement as it does raise the bar significantly.

Reply:
> > Ok, so this should only be used for event publishing, for informing interested parties that something has occurred?
> 
> Correct
> 
> > And this mechanism should not be used to ask another party to do something?
> 
> Also correct
> 
> > And the same event can be delivered to the same subscriber more than once?
> 
> Not in the initial version. Note that doing so would require imposing an idempotency requirement on the subscriber's event handler. It is probably a good requirement anyways, but I would be hesitant to impose it for the initial version as a strict requirement as it does raise the bar significantly.

I agree that requiring idempotency does raise the bar significantly.

Is it reasonable to say that, from the perspective of a subscriber, this initial version is delivery at most once?

Reply:
Yes.